external_libs/ffprobe-53/doc/ffprobe.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "FFPROBE 1"
.TH FFPROBE 1 "2008-03-30" " " " "
.SH "NAME"
ffprobe \- FFmpeg based simple media prober
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
ffprobe [options] \fIinput_files\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
FFprobe is a very simple and portable prober using the FFmpeg
libraries to probe the content of a file. 
.PP
It may be used to print informations about the format and header of a
multimedia file or stream, the multimedia streams and each single
packet or frame.
.PP
FFprobe output consists in stanza of the form:
.PP
.Vb 5
\&        [TAG]
\&        key1=val1
\&        ...
\&        keyN=valN
\&        [/TAG]
.Ve
.PP
and is designed in order to be easily parsable by a textual filter.
.PP
FFprobe may be used both as a standalone application or in combination with
a textual filter, which is supposed to perform more sophisticated
elaboration (e.g. statistics elaboration or plotting).
.SH "OPTIONS"
.IX Header "OPTIONS"
.Sh "Main options"
.IX Subsection "Main options"
.IP "\fB\-h\fR" 4
.IX Item "-h"
show help and exit
.IP "\fB\-version\fR" 4
.IX Item "-version"
show version and exit
.IP "\fB\-L\fR" 4
.IX Item "-L"
show license and exit
.IP "\fB\-pretty\fR" 4
.IX Item "-pretty"
will show every single value using an adequate unit of measure and
using more eye-ball parsable prefixes.
.IP "\fB\-k\fR" 4
.IX Item "-k"
keep going even in case of error, always returns 0
.IP "\fB\-read_packets\fR" 4
.IX Item "-read_packets"
read packets info
.IP "\fB\-read_frames\fR" 4
.IX Item "-read_frames"
read frames info
.IP "\fB\-show_files\fR" 4
.IX Item "-show_files"
show file info, that is the informations relative to the container
.IP "\fB\-show_frames\fR" 4
.IX Item "-show_frames"
show frames info, implies the option \-read_frames and \-read_packets
.IP "\fB\-show_files\fR" 4
.IX Item "-show_files"
show files info, implies the option \-read_packets
.IP "\fB\-show_streams\fR" 4
.IX Item "-show_streams"
show streams info, that is the informations relative to the monomedia
streams contained in the header 
.IP "\fB\-show_tags\fR" 4
.IX Item "-show_tags"
show tags info: actually are supported only the track, title, author, copyright,
comment, album, year, and genre tags.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIffmpeg\fR\|(1), \fIffserver\fR\|(1), \fIffplay\fR\|(1)
.SH "AUTHOR"
.IX Header "AUTHOR"
Stefano Sabatini


lib/ffprobe.js
var spawn = require('child_process').spawn,
		path = require('path');

module.exports = (function() {
	function parseBlock(block) {
		var block_object = {}, lines = block.split('\n');

		lines.forEach(function(line) {
			var data = line.split('=');
			if(data && data.length === 2) block_object[data[0]] = convert(data[1].trim());
		});

		return block_object;
	};

	function convert(str) {
		return str.match(/^\d+\.?\d*$/) ? parseFloat(str) : str;
	};

	function parseStreams(text, callback) {
		var start = text.indexOf('[STREAM]'),
				end		= text.lastIndexOf('[/STREAM]');

		if(start === -1 || end === -1) return { streams: null };

		var streams = [];
		var parsed = text.slice(start + 8, end).trim();
		parsed = parsed.replace('[STREAM]\n', '');

		blocks = parsed.split('[/STREAM]');

		blocks.forEach(function(stream, idx) {
			var codec_data = parseBlock(stream);
			var sindex = codec_data.index;
			delete codec_data.index;

			if(sindex) streams[sindex] = codec_data;
			else streams.push(codec_data);
		});

		return { streams: streams };
	};

	function parseFormat(text, callback) {
		var start = text.indexOf('[FORMAT]'),
				end		= text.lastIndexOf('[/FORMAT]');

		if(start === -1 || end === -1) return { format: null }

		var parsed = text.slice(start + 8, end).trim();
		block = parsed.replace('[FORMAT]\n', '').replace('[/FORMAT]', '');
		var raw_format = parseBlock(block), format = { }, metadata = { };

		//REMOVE metadata
		delete raw_format.filename;
		for(var attr in raw_format) {
			if(raw_format.hasOwnProperty(attr)) {
				if(attr.indexOf('TAG') === -1) format[attr] = raw_format[attr];
				else metadata[attr.slice(4)] = raw_format[attr];
			}
		}

		return { format: format, metadata: metadata };
	};

	return function(file, callback) {
		var proc = spawn('ffprobe', ['-show_streams', '-show_format', '-convert_tags', file]), probeData = []
		var start = Date.now();

		proc.stdout.on('data', function(data) { probeData.push(data.toString()) });

		proc.on('exit', function() {
			var pdata = probeData.join('');
			var s = parseStreams(pdata), f = parseFormat(pdata);

			callback({
				filename: path.basename(file),
				filepath: path.dirname(file),
				fileext: path.extname(file),
				file: file,
				probe_time: Date.now() - start,
				streams: s.streams,
				format: f.format,
				metadata: f.metadata
			});
		});
	};
})()
